Makefile

```diff
@@ -1,46 +1,110 @@
+include .make
 
-BRANCH?=$(shell git rev-parse --abbrev-ref HEAD)
+ANSIBLE_INSTALL_VERSION ?= 2.2.3.0
+ANSIBLE_CONFIG ?= tests/ansible.cfg
+ROLE_NAME ?= $(shell basename $$(pwd))
+TEST_PLAYBOOK ?= test.yml
+VAGRANT_BOX ?= ubuntu/trusty64
+PATH := $(PWD)/.venv_$(ANSIBLE_INSTALL_VERSION)/bin:$(shell printenv PATH)
+SHELL := env PATH=$(PATH) /bin/bash
+
+.DEFAULT_GOAL := help
+.PHONY: help
+
+export ANSIBLE_CONFIG
+export PATH
+export VAGRANT_BOX
+export TEST_PLAYBOOK
 
 all: test clean
 
+## Run tests on any file change
 watch: test_deps
 	while sleep 1; do \
-		find defaults/ handlers/ meta/ tasks/ templates/ tests/vagrant/test.yml \
-		| entr -d make lint vagrant_up; \
+		find defaults/ handlers/ meta/ tasks/ templates/ tests/test.yml \
+		| entr -d make lint vagrant; \
 	done
 
-test: lint test_deps vagrant_up
+## Run tests
+test: lint test_deps vagrant
+
+## Install test dependencies
+test_deps: .venv_$(ANSIBLE_INSTALL_VERSION) tests/roles
+
+tests/roles:
+	mkdir -p tests/roles
+	ln -s ../.. tests/roles/sansible.$(ROLE_NAME)
+
+## ! Executes Ansible tests using local connection
+# run it ONLY from within a test VM.
+# If you want to test this role, run `make test` instead.
+# Example: make test_ansible
+#          make test_ansible TEST_PLAYBOOK=test-something-else.yml
+test_ansible: test_ansible_build test_ansible_configure
+	cd tests && ansible-playbook \
+		--inventory inventory \
+		--connection local \
+		--tags assert \
+		$(TEST_PLAYBOOK)
+
+## ! Executes Ansible tests using local connection
+# run it ONLY from witinh a test VM.
+# Example: make test_ansible_build
+#          make test_ansible_build TEST_PLAYBOOK=test-something-else.yml
+test_ansible_%:
+	cd tests && ansible-playbook \
+		--inventory inventory \
+		--connection local \
+		--tags=$(subst test_ansible_,,$@) \
+		$(TEST_PLAYBOOK)
+	cd tests && ansible-playbook \
+		--inventory inventory \
+		--connection local \
+		--tags=$(subst test_ansible_,,$@) \
+		$(TEST_PLAYBOOK) \
+		| grep -q 'changed=0.*failed=0' \
+			&& (echo 'Idempotence test: pass' && exit 0) \
+			|| (echo 'Idempotence test: fail' && exit 1)
 
-integration_test: clean integration_test_deps vagrant_up clean
+## Start and (re)provisiom Vagrant test box
+vagrant:
+	cd tests && vagrant up --no-provision
+	cd tests && vagrant provision
+	@echo "- - - - - - - - - - - - - - - - - - - - - - -"
+	@echo "           Provisioning Successful"
+	@echo "- - - - - - - - - - - - - - - - - - - - - - -"
 
-test_deps:
-	rm -rf tests/vagrant/sansible.*
-	ln -s ../.. tests/vagrant/sansible.ansible
+## Execute simple Vagrant command
+# Example: make vagrant_ssh
+#          make vagrant_halt
+vagrant_%:
+	cd tests && vagrant $(subst vagrant_,,$@)
 
-integration_test_deps:
-	sed -i.bak \
-		-E 's/(.*)version: (.*)/\1version: origin\/$(BRANCH)/' \
-		tests/vagrant/integration_requirements.yml
-	rm -rf tests/vagrant/sansible.*
+## Installs a virtual environment and all python dependencies
+.venv_%:
```

Comment:
This seems weird.. why accept a parameter in the target name, but then proceed to ignore the parameter?

Reply:
I think the idea here is to keep this target Ansible version agnostic.

Reply:
It could just be called ".venv" in that case :)

Reply:
Check the rest of the Makefile, that wouldn't catch calls like this:

```
make .venv_2.4.1.0
```

The idea I think (bearing in mind I didn't write the code :) ) is to enable easy testing of multiple Ansible versions.

Reply:
Actually.. make checks if a file/folder of the target name exists to decide if the target needs to be processed or not.. doesnt it? In which case the wildcard in the target is appropriate.
Not a big deal. :)

Reply:
Yep that's correct, so yeah I think this format makes sense (I blame Wojtek if not).